File: CSharp/PaymentsService/Infrastructure/Payments/CreditCard/CreditCardProcessPaymentRequestValidator.cs
using PaymentsService.Domain.Payments;

namespace PaymentsService.Infrastructure.Payments.CreditCard;

public sealed class CreditCardProcessPaymentRequestValidator : IProcessPaymentRequestValidator
{
    private CreditCardProcessPaymentRequestValidator() { }


    private static readonly Lazy<CreditCardProcessPaymentRequestValidator> _lazyInstance = new(() => new CreditCardProcessPaymentRequestValidator());
    public static CreditCardProcessPaymentRequestValidator Instance => _lazyInstance.Value;


    public RequestValidationResult ValidateRequest(ProcessPaymentBaseRequest request)
    {
        if (request is not ProcessCreditCardPaymentRequest req || req.CardInfo is null)
        {
            return new RequestValidationResult(false, "BadRequest");
        }

        List<string> errors = new(7);

        if (string.IsNullOrEmpty(req.CardInfo.FirstName))
        {
            errors.Add("BadFirstName");
        }

        if (string.IsNullOrEmpty(req.CardInfo.LastName))
        {
            errors.Add("BadLastName");
        }

        if (string.IsNullOrEmpty(req.CardInfo.Type))
        {
            errors.Add("BadType");
        }

        if (string.IsNullOrEmpty(req.CardInfo.Number) || req.CardInfo.Number.Length < 12)
        {
            errors.Add("BadNumber");
        }

        if (req.CardInfo.CVV < 100 || req.CardInfo.CVV > 1000)
        {
            errors.Add("BadCVV");
        }

        if(req.CardInfo.ExpYear < DateTime.UtcNow.Year)
        {
            errors.Add("BadExp");
        }
        else if (req.CardInfo.ExpYear == DateTime.UtcNow.Year && req.CardInfo.ExpMonth < DateTime.UtcNow.Month)
        {
            errors.Add("BadExp");
        }

        if(errors.Any())
        {
            return new RequestValidationResult(false, string.Join(", ", errors));
        }

        return new RequestValidationResult(true);
    }
}

File: Python/PaymentsService/Infrastructure/Payments/CreditCard/CreditCardProcessPaymentRequestValidator.py
from datetime import datetime

from Domain.Payments.IProcessPaymentRequestValidator import IProcessPaymentRequestValidator
from Domain.Payments.RequestValidationResult import RequestValidationResult
from Infrastructure.Payments.CreditCard.ProcessCreditCardPaymentRequest import ProcessCreditCardPaymentRequest

class CreditCardProcessPaymentRequestValidator(IProcessPaymentRequestValidator):
    def __init__(self):
        pass

    def validate_request(self, request):
        if not isinstance(request, ProcessCreditCardPaymentRequest) or request.cardInfo is None:
            return RequestValidationResult(False, "BadRequest")

        errors = []

        if not request.cardInfo.firstName:
            errors.append("BadFirstName")

        if not request.cardInfo.lastName:
            errors.append("BadLastName")

        if not request.cardInfo.type:
            errors.append("BadType")

        if not request.cardInfo.number or len(request.cardInfo.number) < 12:
            errors.append("BadNumber")

        if request.cardInfo.cvv < 100 or request.cardInfo.cvv > 1000:
            errors.append("BadCVV")

        current_year = datetime.utcnow().year

        if request.cardInfo.expYear < current_year:
            errors.append("BadExp")
        elif request.cardInfo.expYear == current_year and request.cardInfo.expMonth < datetime.utcnow().month:
            errors.append("BadExp")

        if errors:
            return RequestValidationResult(False, ", ".join(errors))

        return RequestValidationResult(True)